Error applying plan for proxmox_virtual_environment_storage_directory

**Describe the bug**

With the following resource definition

```
resource "proxmox_virtual_environment_storage_directory" "this" {
  id   = "local"
  path = "/var/lib/vz"

  content = ["backup", "iso", "snippets", "vztmpl"]
  shared  = false
  disable = false
}
```

I run into the following error

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with
the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # proxmox_virtual_environment_storage_directory.this will be updated in-place
  ~ resource "proxmox_virtual_environment_storage_directory" "this" {
        id      = "local"
      ~ nodes   = [] -> (known after apply)
        # (4 unchanged attributes hidden)

      - backups {
          - keep_all = true -> null
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
cd proxmox
TF_IN_AUTOMATION=1 TF_CLI_ARGS=-no-color terraform apply tfplan-proxmox
proxmox_virtual_environment_storage_directory.this: Modifying... [id=local]

Error: Provider produced inconsistent result after apply

When applying changes to proxmox_virtual_environment_storage_directory.this, provider
"provider[\"registry.terraform.io/bpg/proxmox\"]" produced an unexpected new value: .backups: was absent, but now
present.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
make: *** [Makefile:30: proxmox-apply] Error 1
```

**Expected behavior**
No error on applying the plan


- Single or clustered Proxmox: cluster of 3 nodes
- Proxmox version: 9.1.4
- Provider version (ideally it should be the latest version): 0.91.0
- Terraform/OpenTofu version: v1.14.3
- OS (where you run Terraform/OpenTofu from): Linux (amd64)

--- a/fwprovider/storage/model_cifs.go
+++ b/fwprovider/storage/model_cifs.go
@@ -117,16 +117,11 @@ func (m *CIFSStorageModel) fromAPI(ctx context.Context, datastore *storage.Datas
 		m.SnapshotsAsVolumeChain = types.BoolValue(*datastore.SnapshotsAsVolumeChain.PointerBool())
 	}
 
-	if datastore.MaxProtectedBackups != nil || (datastore.PruneBackups != nil && *datastore.PruneBackups != "") {
-		if m.Backups == nil {
-			m.Backups = &BackupModel{}
-		}
-
+	// only populate backups if user has configured it to avoid "was absent, but now present" error
+	if m.Backups != nil {
 		if err := m.Backups.fromAPI(datastore.MaxProtectedBackups, datastore.PruneBackups); err != nil {
 			return err
 		}
-	} else {
-		m.Backups = nil
 	}
 
 	return nil

--- a/fwprovider/storage/model_directory.go
+++ b/fwprovider/storage/model_directory.go
@@ -87,16 +87,11 @@ func (m *DirectoryStorageModel) fromAPI(ctx context.Context, datastore *storage.
 		m.Preallocation = types.StringValue(*datastore.Preallocation)
 	}
 
-	if datastore.MaxProtectedBackups != nil || (datastore.PruneBackups != nil && *datastore.PruneBackups != "") {
-		if m.Backups == nil {
-			m.Backups = &BackupModel{}
-		}
-
+	// only populate backups if user has configured it to avoid "was absent, but now present" error
+	if m.Backups != nil {
 		if err := m.Backups.fromAPI(datastore.MaxProtectedBackups, datastore.PruneBackups); err != nil {
 			return err
 		}
-	} else {
-		m.Backups = nil
 	}
 
 	return nil

--- a/fwprovider/storage/model_nfs.go
+++ b/fwprovider/storage/model_nfs.go
@@ -103,16 +103,11 @@ func (m *NFSStorageModel) fromAPI(ctx context.Context, datastore *storage.Datast
 		m.SnapshotsAsVolumeChain = types.BoolValue(*datastore.SnapshotsAsVolumeChain.PointerBool())
 	}
 
-	if datastore.MaxProtectedBackups != nil || (datastore.PruneBackups != nil && *datastore.PruneBackups != "") {
-		if m.Backups == nil {
-			m.Backups = &BackupModel{}
-		}
-
+	// only populate backups if user has configured it to avoid "was absent, but now present" error
+	if m.Backups != nil {
 		if err := m.Backups.fromAPI(datastore.MaxProtectedBackups, datastore.PruneBackups); err != nil {
 			return err
 		}
-	} else {
-		m.Backups = nil
 	}
 
 	return nil

--- a/fwprovider/storage/model_pbs.go
+++ b/fwprovider/storage/model_pbs.go
@@ -123,16 +123,11 @@ func (m *PBSStorageModel) fromAPI(ctx context.Context, datastore *storage.Datast
 		m.Shared = types.BoolValue(*datastore.Shared.PointerBool())
 	}
 
-	if datastore.MaxProtectedBackups != nil || (datastore.PruneBackups != nil && *datastore.PruneBackups != "") {
-		if m.Backups == nil {
-			m.Backups = &BackupModel{}
-		}
-
+	// only populate backups if user has configured it to avoid "was absent, but now present" error
+	if m.Backups != nil {
 		if err := m.Backups.fromAPI(datastore.MaxProtectedBackups, datastore.PruneBackups); err != nil {
 			return err
 		}
-	} else {
-		m.Backups = nil
 	}
 
 	return nil

--- a/fwprovider/test/resource_storage_directory_test.go
+++ b/fwprovider/test/resource_storage_directory_test.go
@@ -253,3 +253,70 @@ func TestAccResourceStorageDirectory(t *testing.T) {
 		},
 	})
 }
+
+// TestAccResourceStorageDirectoryRemoveBackups tests removing backups block from storage directory.
+// Regression test for https://github.com/bpg/terraform-provider-proxmox/issues/2463
+func TestAccResourceStorageDirectoryRemoveBackups(t *testing.T) {
+	te := InitEnvironment(t)
+
+	storageID := fmt.Sprintf("dir-%s-%d", strings.ToLower(gofakeit.Word()), time.Now().UnixNano())
+	dirPath := utils.GetAnyStringEnv("PROXMOX_VE_ACC_STORAGE_DIR_PATH")
+	if dirPath == "" {
+		dirPath = "/var/lib/vz"
+	}
+
+	te.AddTemplateVars(map[string]any{
+		"StorageID": storageID,
+		"DirPath":   dirPath,
+	})
+
+	resource.ParallelTest(t, resource.TestCase{
+		ProtoV6ProviderFactories: te.AccProviders,
+		Steps: []resource.TestStep{
+			// create with backups block
+			{
+				Config: te.RenderConfig(`
+				resource "proxmox_virtual_environment_storage_directory" "test" {
+					id      = "{{.StorageID}}"
+					path    = "{{.DirPath}}"
+					content = ["backup", "iso", "snippets", "vztmpl"]
+					shared  = false
+					disable = false
+
+					backups {
+						keep_all = true
+					}
+				}`),
+				Check: resource.ComposeTestCheckFunc(
+					ResourceAttributes("proxmox_virtual_environment_storage_directory.test", map[string]string{
+						"id":               storageID,
+						"path":             dirPath,
+						"shared":           "false",
+						"disable":          "false",
+						"backups.keep_all": "true",
+					}),
+				),
+			},
+			// remove backups block - should not cause "was absent, but now present" error
+			{
+				Config: te.RenderConfig(`
+				resource "proxmox_virtual_environment_storage_directory" "test" {
+					id      = "{{.StorageID}}"
+					path    = "{{.DirPath}}"
+					content = ["backup", "iso", "snippets", "vztmpl"]
+					shared  = false
+					disable = false
+				}`),
+				Check: resource.ComposeTestCheckFunc(
+					ResourceAttributes("proxmox_virtual_environment_storage_directory.test", map[string]string{
+						"id": storageID,
+					}),
+					resource.TestCheckNoResourceAttr(
+						"proxmox_virtual_environment_storage_directory.test",
+						"backups",
+					),
+				),
+			},
+		},
+	})
+}

